Users/taosang/SWDEV-510994 - Refactor atomics header and tests

TomSang opened the pull request:
## Motivation

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

    1. Introduce __HIP_ATOMIC_BACKWARD_COMPAT.
    By default we define __HIP_ATOMIC_BACKWARD_COMPAT=1 to
    let hip atomic functions maintain old assumptions. if
    users want to adopt the new behavior, that is , by default
    assume no-fine-grained no-remote-memory, then they can
    define __HIP_ATOMIC_BACKWARD_COMPAT=0 and get the new
    behaviour.

    2. Use  __HIP_ATOMIC_BACKWARD_COMPAT_MEMORY to replace
    original __HIP_FINE_GRAINED_MEMORY  in atomic header.
    And apply __HIP_ATOMIC_BACKWARD_COMPAT_MEMORY onto all 
    atomicXXX_system() functions to prevent failure on memory
    allocated by hipHostMalloc().

    3. Replace HIP_TEST_FINE_GRAINED_MEMORY with
    HIP_TEST_ATOMIC_BACKWARD_COMPAT_MEMORY in hip-tests.
  
    4. Fix negative test errors.
        Fix managed memory test error on memory order.
        some other minor changes.
        As a result  all originally disabled tests are enabled.
   
    5. Add more atomics tests in some cases.
    
    6. Reduce test time in each case.
         Reduce iteration number to 1 for tests that cost too much time.
    
    8. Put common codes into hip_test_common.hh
      

<!-- Explain the changes along with any relevant GitHub links. -->

Atomics header and hip-tests

<!-- Explain any relevant testing done to verify this PR. -->

Pass

<!-- Briefly summarize test outcomes. -->

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

TomSang committed:
SWDEV-550626 - Refactor atomics header and tests

1. Fix some system scoped atomics test failures by
introducing __HIP_ATOMIC_BACKWARD_COMPAT
and __HIP_ATOMIC_BACKWARD_COMPAT_MEMORY which replaces
original __HIP_FINE_GRAINED_MEMORY.

By default we define __HIP_ATOMIC_BACKWARD_COMPAT=1 to
let hipAtomic functions maintain old assumptions. if
users want to adopt the new behavior, that is , by default
assume no-fine-grained no-remote-memory, then they can
define __HIP_ATOMIC_BACKWARD_COMPAT=0 and get the new
behavior.

2. Fix some wrong logics in p2p atomics tests.
   Fix negative test errors.
   Fix managed memory test error on memory order.

3. Add more atomics test cases.

4. Reduce test time in each case.

5. Put common codes into hip_test_common.hh

TomSang committed:
SWDEV-510994 - Some improvements

1. Replace __HIP_FINE_GRAINED_MEMORY with
__HIP_ATOMIC_BACKWARD_COMPAT_MEMORY for some functions
in float types.

2. Replace HIP_TEST_FINE_GRAINED_MEMORY with
HIP_TEST_ATOMIC_BACKWARD_COMPAT_MEMORY in hip-tests.

3. Remove P2P logics that leads to sporadic failures in stressed
tests.

4. Reduce interation number to 1 in all system tests to save time.

yxsamliu commented on projects/hip-tests/catch/unit/atomics/arithmetic_common.hh:
```diff
@@ -122,11 +116,21 @@ __device__ TestType PerformAtomicOperation(TestType* const mem, const LinearAllo
   } else if constexpr (operation == AtomicOperation::kAddSystem) {
-    return atomicAdd_system(mem, val);
+    if (allocType == LinearAllocs::hipHostMalloc)
+      HIP_TEST_ATOMIC_BACKWARD_COMPAT_MEMORY {
```
HIP_TEST_ATOMIC_BACKWARD_COMPAT_MEMORY has no effect here since [[clang::atomic]] only affects clang atomic builtins and atomic expressions. it does not affect ordinary functions like atomicAdd_system. Better remove it to avoid confusing the users. same as below

TomSang replied:
I tried to remove HIP_TEST_ATOMIC_BACKWARD_COMPAT_MEMORY for kAddSystem,   kSubSystem and kCASAddSystem , then there are errors due to this as follows,

99% tests passed, 8 tests failed out of 609

Total Test time (real) = 1906.66 sec

The following tests FAILED:
	388 - Unit_atomicAdd_system_Positive_Host_And_Peer_GPUs - int (Failed)
	392 - Unit_atomicAdd_system_Positive_Host_And_Peer_GPUs - float (Failed)
	423 - Unit_atomicSub_system_Positive_Peer_GPUs - float (Failed)
	435 - Unit_atomicSub_system_Positive_Host_And_Peer_GPUs - float (Failed)
	446 - Unit_atomicCAS_system_Positive_Peer_GPUs - int (Failed)
	449 - Unit_atomicCAS_system_Positive_Peer_GPUs - unsigned short int (Failed)
	456 - Unit_atomicCAS_system_Positive_Host_And_Peer_GPUs - unsigned long long (Failed)
	457 - Unit_atomicCAS_system_Positive_Host_And_Peer_GPUs - unsigned short int (Failed)
Errors while running CTest


Next round test,
99% tests passed, 2 tests failed out of 609

Total Test time (real) = 1978.67 sec

The following tests FAILED:
	381 - Unit_atomicAdd_system_Positive_Peer_GPUs - double (Failed)
	391 - Unit_atomicAdd_system_Positive_Host_And_Peer_GPUs - unsigned long long (Failed)

TomSang committed:
SWDEV-510994 - Remove HIP_TEST_ATOMIC_BACKWARD_COMPAT_MEMORY from some
system function callings.

TomSang committed:
SWDEV-510994 - Workarround new compiler issue with atomicXXX_system()

Add __HIP_ATOMIC_BACKWARD_COMPAT_MEMORY to all atomicXXX_system()
functions in amd_hip_atomic.h to workarround new compiler issue on
memory allocated by hipHostMall().

TomSang committed:
Merge branch 'develop' into users/tomsang/SWDEV-550626-workarround-atomics-tests-issue-with-new-compiler

TomSang committed:
Merge branch 'develop' into users/tomsang/SWDEV-550626-workarround-atomics-tests-issue-with-new-compiler

TomSang committed:
Merge branch 'develop' into users/tomsang/SWDEV-550626-workarround-atomics-tests-issue-with-new-compiler

TomSang committed:
Merge branch 'develop' into users/tomsang/SWDEV-550626-workarround-atomics-tests-issue-with-new-compiler

TomSang committed:
Merge branch 'develop' into users/tomsang/SWDEV-550626-workarround-atomics-tests-issue-with-new-compiler
